### PR TITLE
Fix issue 15421 - The behaviours of the topNs differ with the bottom

### PR DESCRIPTION
**Pull request by Infiltrator**
The behaviour with regard to the bottoms or right ranges of the topNs
differed in that topN(R, i) correctly moved elements around to preserve
each element, but that topN(R, R) did not and ended up with duplicate
and missing elements.

### CODE REVIEW

File: std/algorithm/sorting.d [hunk 1]
```diff
@@ -2294,9 +2294,15 @@ void topN(alias less = "a < b",
...
     auto heap = BinaryHeap!Range1(r1);
+    auto shadow = r2;
+    size_t i = 0;
     for (; !r2.empty; r2.popFront())
     {
-        heap.conditionalInsert(r2.front);
+        auto f = heap.front;
+        if (heap.conditionalInsert(r2.front))
+        {
+           shadow[i++] = f;
```
Comment (Infiltrator) [inline std/algorithm/sorting.d]: Should I use lockstep here instead of indexing?

File: std/algorithm/sorting.d [hunk 2]
```diff
@@ -2294,9 +2294,15 @@ void topN(alias less = "a < b",
     static assert(ss == SwapStrategy.unstable,
             "Stable topN not yet implemented");
     auto heap = BinaryHeap!Range1(r1);
+    auto shadow = r2;
```
Comment (JackStouffer) [inline std/algorithm/sorting.d]: `r2.save`

File: std/algorithm/sorting.d [hunk 3]
```diff
@@ -2294,9 +2294,15 @@ void topN(alias less = "a < b",
...
+    size_t i = 0;
     for (; !r2.empty; r2.popFront())
```
Comment (JackStouffer) [inline std/algorithm/sorting.d]: This can be refactored to use `foreach` easily, so might as well

File: std/algorithm/sorting.d (hunk 1 above)
Comment (JackStouffer) [inline std/algorithm/sorting.d]: Yes, in combination with a `foreach` and passing by `ref`

**Comment by JackStouffer** [comment]
I understand that this is just supposed to be a bug fix for 15421, but the fact that the only ranges that are tested are regular arrays is worrisome when this changes internal behavior. Please add a `unittest` that tests with the ranges in `std.internal.testing.dummyrange`

File: std/algorithm/sorting.d [hunk 4]
```diff
@@ -2310,6 +2316,23 @@ unittest
...
+// bug 15421
+unittest
+{
+    auto a = [ 9, 8, 0, 3, 5, 25, 43, 4, 2, 0, 7 ];
+    auto b = [ 9, 8, 0, 3, 5, 25, 43, 4, 2, 0, 7 ];
+
+    topN(a, 4);
+    topN(b[0 .. 4], b[4 .. $]);
+
+    sort(a[0 .. 4]); sort(a[4 .. $]);
```
Comment (andralex) [inline std/algorithm/sorting.d]: one stmt per line please

**Comment by andralex** [comment]
plus @JackStouffer's comments. Thanks!

File: std/algorithm/sorting.d (hunk 3 above)
Comment (Infiltrator) [inline std/algorithm/sorting.d]: Done.

File: std/algorithm/sorting.d (hunk 4 above)
Comment (Infiltrator) [inline std/algorithm/sorting.d]: Done.

**Comment by Infiltrator** [comment]
> I understand that this is just supposed to be a bug fix for 15421, but the fact that the only ranges that are tested are regular arrays is worrisome when this changes internal behavior.

I agree with you and I'm all for testing.  One of the reasons that I like D so much is how easy it makes testing.

> Please add a unittest that tests with the ranges in std.internal.testing.dummyrange

So do you want me to just make a unittest for each of the four Reference*Range ranges?

File: std/algorithm/sorting.d (hunk 3 above)
Comment (JackStouffer) [inline std/algorithm/sorting.d]: I don't see it, did you push?

**Comment by JackStouffer** [comment]
> So do you want me to just make a unittest for each of the four Reference*Range ranges?

No, it's better that each `unittest` block contains all of the tests of one type. Something like this, this should go above the `unittest` block for the bug

```
unittest
{
    import std.internal.testing.dummyrange;
    alias RandomRanges = AliasSeq!(
        DummyRange!(ReturnBy.Reference, Length.Yes, RangeType.Random),
        DummyRange!(ReturnBy.Value, Length.Yes, RangeType.Random),
    );
    foreach (T1; RandomRanges)
    {
        foreach (T2; AllDummyRanges) 
        {
            // tests
        }
    }
}
```

This way you test every type of random access range with every type of range.

File: std/algorithm/sorting.d (hunk 3 above)
Comment (Infiltrator) [inline std/algorithm/sorting.d]: No, sorry.  I was going to push it together with the unittests, but I haven't got those right yet.  In hindsight, I probably shouldn't have commented until I had pushed.

**Comment by Infiltrator** [comment]
I must be doing something wrong here, but I can't get even a very simple unittest going with all the dummy ranges.  Here is what I have boiled it down to:

    unittest
    {
        import std.internal.test.dummyrange;
        foreach (T; AllDummyRanges)
        {
            T a;
            T b;
            topN(a, b);
        }
    }

But I can't get that to compile.

**Comment by JackStouffer** [comment]
Because the constraints require the first argument to be a random access range with a length. The first range is the AllDummyRanges AliasSeq is a forward range.

**Comment by Infiltrator** [comment]
Right, I was probably too eager with trying to reduce the error.  But even using only the two random ranges with length as below gives opSlicing errors even without the changes in this PR.

```d
unittest
{
    import std.internal.test.dummyrange;
	import std.meta : AliasSeq;
	alias RandomRanges = AliasSeq!(
        DummyRange!(ReturnBy.Reference, Length.Yes, RangeType.Random),
        DummyRange!(ReturnBy.Value, Length.Yes, RangeType.Random),
    );
    foreach (T; RandomRanges)
    {
        T a;
        T b;
        topN(a, b);
    }
}
```

    std/container/binaryheap.d(150): Error: function std.internal.test.dummyrange.DummyRange!(cast(ReturnBy)0, cast(Length)0, cast(RangeType)3).DummyRange.opSlice (ulong lower, ulong upper) is not callable using argument types ()
Plus a few dozen more lines.  I don't want to paste them all here so you can run the code yourself if you want to see.

BinaryHeap.acquire tries to call `buildHeap(s[])` and DummyRange does not like that (it has no overload for opSlice(); it has only opSplice(lower, upper)).  Should BinaryHeap support ranges without an opSlice()?  Maybe using s.save instead?

**Comment by JakobOvrum** [comment]
`r[]` is not a range primitive, it just happens to work with slices which is the only type it's tested against. It should be:
```d
static if (isRandomAccessRange!Store)
    buildHeap(s);
else
    buildHeap(s[]); // `s` is a container
```

**Comment by Infiltrator** [comment]
Okay, here's what I've got so far.  I'm not sure what to do with the dup, but I'll look at that and sort after I get some sleep.

File: std/container/binaryheap.d [hunk 5]
```diff
@@ -115,11 +115,26 @@ if (isRandomAccessRange!(Store) || isRandomAccessRange!(typeof(Store.init[])))
     {
         assert(!store.empty, "Cannot pop an empty store.");
         if (store.length == 1) return;
-        auto t1 = moveFront(store[]);
-        auto t2 = moveBack(store[]);
+        static if(isRandomAccessRange!Store)
+        {
+            auto t1 = moveFront(store);
```
Comment (JakobOvrum) [inline std/container/binaryheap.d]: Perhaps instead of using `store` everywhere, add an internal primitive like:

```d
static if (isRandomAccessRange!Store)
    alias range = store;
else
    auto range() @property
    {
        return store[];
    }
```
Then use `range` everywhere. However, `BinaryHeap` clearly needs more tests too, and at this point, it would probably be better to fix it in a separate PR. It also has a strange thing going on where it uses the `package(std)` symbol `HeapOps` from std.algorithm.sorting.

Comment (Infiltrator) [inline std/container/binaryheap.d]: Yeah, I was thinking something like that.  But then there's functions like pop(Store store).  I guess that I could just leave it be.

**Comment by JackStouffer** [comment]
ping

**Comment by wilzbach** [comment]
ping @Infiltrator - do you have time to rebase?
